docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,235 @@
+  * Adding a fourth component, the"orphan" finalizer:
+    * Watches for update events as described in [2A](#new-components-1).
+    * Removes the object in the event from the `ParentReferences` of its children.
+      * children can be found via the DAG kept by the GC, or by relisting the child resource and checking the ParentReferences field of each potential child object.
+    * Removes the finalizer from the `ObjectMeta.Finalizers` of this object.
+
+# Comparison between design #1 and #2
+
+The main difference between the two designs is when to update the ParentReferences. In design #1, because a tombstone is created to indicate "orphaning" is desired, the updates to ParentReferences can be deferred until the deletion of the tombstone. In design #2, the updates need to be done before the parent object is deleted from the registry.
+
+* Advantages of "Tombstone + GC" design
+  * Faster to free the resource name compared to using finalizers. The original object can be deleted to free the resource name once the tombstone is created, rather than waiting for the finalizers to update all children’s ObejectMeta.ParentReferences.
+* Advantages of "Finalizer Framework + GC"
+  * The finalizer framework is needed for other purposes as well.
+
+# Common designs
+
+## Orphan adoption
+
+Controllers are responsible for adopting orphaned child resources. To do so, controllers
+* Checks a potential child object’s ParentReferences to determine if it is orphaned.
+* Fills the ParentReferences if the object matches the controller’s selector and is orphaned.
+
+There is a potential race between the GC removing a parent reference and the controllers adding it back during adoption. Imaging this case: a user deletes a parent object and intends to orphan the child objects, so the GC removes the parent from the child object's ParentReferences list, but the controller of the parent resource hasn't observed the deletion yet, so it adopts the child again and adds the reference back.
+* In design #1 (the tombstone one), this race does not matter if the tombstone lives long enough to prevent the child to be garbage collected.
+* In design #2, this race does matter, because the child will be mistakenly garbage collected. This race can be avoided by implementing Status.ObservedGeneration in all resources. Before updating the child Object's ParentReferences, the "orphan" finalizer checks Status.ObservedGeneration of the parent object to ensure its controller has already observed the deletion.
+
+## Upgrading a cluster to support cascading deletion
+
+For the master, after upgrading to a version that supports cascading deletion, the ParentReferences of existing objects remain empty, so the controllers will regard them as orphaned and start the adoption procedures. After the adoptions are done, server-side cascading will be effective for these existing objects.
```

Comment:
New resources/versions can have new default behavior.

Existing resources/versions could introduce cascading deletion as a non-default option.

Setting parent references for existing resources will be non-trivial, but I don't see a way to avoid it, given that controllers can adopt existing resources.

We may need to keep the kubectl deletion logic indefinitely.